File: Client/Android/app/src/main/java/com/vertcdemo/app/SceneEntryFragment.java
// Copyright (c) 2023 BytePlus Pte. Ltd.
// SPDX-License-Identifier: Apache-2.0

package com.vertcdemo.app;

import android.content.Context;
import android.graphics.drawable.AnimationDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.core.content.res.ResourcesCompat;
import androidx.fragment.app.Fragment;

import com.vertcdemo.app.databinding.ItemSceneEntryBinding;
import com.vertcdemo.core.utils.AppUtil;
import com.videoone.app.protocol.SceneEntry;

public class SceneEntryFragment extends Fragment {
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_scene_entry, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        Context context = requireContext();
        ViewGroup cards = view.findViewById(R.id.cards);
        LayoutInflater inflater = LayoutInflater.from(context);

        SceneEntry.forEach(entry -> {
            int title = entry.title();
            int iconRes = entry.icon();
            if (title == ResourcesCompat.ID_NULL || iconRes == ResourcesCompat.ID_NULL) {
                return;
            }

            final ItemSceneEntryBinding binding = ItemSceneEntryBinding.inflate(inflater, cards, false);
            binding.newLabel.setVisibility(entry.isNew() ? View.VISIBLE : View.GONE);
            binding.title.setText(title);

            Drawable icon = ContextCompat.getDrawable(context, entry.icon());
            binding.icon.setImageDrawable(icon);
            if (icon instanceof AnimationDrawable) {
                ((AnimationDrawable) icon).start();
            }

            int description = entry.description();
            if (description != ResourcesCompat.ID_NULL) {
                binding.description.setText(description);
            }

            int action = entry.action();
            if (action != ResourcesCompat.ID_NULL) {
                binding.action.setText(action);
            }

            View card = binding.getRoot();
            card.setOnClickListener(v -> entry.startup(requireActivity()));
            cards.addView(card);
        });
    }
}

File: Client/Android/app/src/main/java/com/vertcdemo/app/SceneEntryFragment.kt
// Copyright (c) 2023 BytePlus Pte. Ltd.
// SPDX-License-Identifier: Apache-2.0
package com.vertcdemo.app

import android.content.Context
import android.graphics.Rect
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.constraintlayout.widget.Guideline
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.vertcdemo.app.databinding.ItemSceneEntryBinding
import com.videoone.app.protocol.SceneEntry

class SceneEntryFragment : Fragment(R.layout.fragment_scene_entry) {

    private lateinit var mMainViewModel: MainViewModel
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mMainViewModel = ViewModelProvider(requireParentFragment())[MainViewModel::class.java]
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val guidelineTop = view.findViewById<Guideline>(R.id.guideline_top)
        mMainViewModel.guidelineTop.observe(viewLifecycleOwner) { top ->
            guidelineTop.setGuidelineBegin(top)
        }

        val context = requireContext()

        view.findViewById<RecyclerView>(R.id.recycler)
            .apply {
                layoutManager = LinearLayoutManager(context)
                addItemDecoration(SpaceItemDecoration(resources.getDimensionPixelSize(R.dimen.item_function_spacing)))
                adapter = SceneItemAdapter(context, SceneEntry.entries)
            }
    }

    class SceneItemHolder(private val binding: ItemSceneEntryBinding) :
        RecyclerView.ViewHolder(binding.root) {
        fun bind(entry: SceneEntry) {
            val context = itemView.context
            binding.title.setText(entry.title)
            binding.description.setText(entry.description)
            binding.background.setImageResource(entry.background)
            if (entry.showGear) {
                binding.gear.visibility = View.VISIBLE
                binding.gear.setOnClickListener { entry.startSettings(context) }
            } else {
                binding.gear.visibility = View.GONE
            }
            binding.root.setOnClickListener { entry.startup(context) }
        }
    }

    class SceneItemAdapter(context: Context, private val entries: List<SceneEntry>) :
        RecyclerView.Adapter<SceneItemHolder>() {
        private val layoutInflater = LayoutInflater.from(context)
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
            SceneItemHolder(ItemSceneEntryBinding.inflate(layoutInflater, parent, false))


        override fun getItemCount(): Int = entries.size

        override fun onBindViewHolder(holder: SceneItemHolder, position: Int) =
            holder.bind(entries[position])
    }

    class SpaceItemDecoration(private val padding: Int) : RecyclerView.ItemDecoration() {
        override fun getItemOffsets(
            outRect: Rect,
            view: View,
            parent: RecyclerView,
            state: RecyclerView.State
        ) {
            val position = parent.getChildLayoutPosition(view)
            outRect.top = if (position == 0) 0 else padding
        }
    }
}